torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,806 @@
+        return {key: self._running_stats[key] / count for key, count in self._running_counts.items()}
+
+
+DEFAULT_MUZERO_TRAINING_CONFIG = MuZeroTrainingConfig(
+    model_publication_interval=500,
+    trial_count=1000,
+    discount_rate=0.99,
+    learning_rate=1e-4,
+    weight_decay=1e-3,
+    bootstrap_steps=20,
+    hidden_dim=128,
+    hidden_layers=2,
+    projector_hidden_dim=128,
+    projector_hidden_layers=1,
+    projector_dim=64,
+    mcts_depth=3,
+    mcts_samples=8,
+    ucb_c1=1.25,
+    ucb_c2=10000.0,
+    batch_size=16,
+    exploration_alpha=0.5,
+    exploration_epsilon=0.25,
+    rollout_length=2,
+    rmin=0.0,  # proto default value issue
+    rmax=0.0,  # proto default value issue
+    vmin=0.0,  # proto default value issue
+    vmax=0.0,  # proto default value issue
+    rbins=16,
+    vbins=16,
+    max_parallel_trials=4,
+    mcts_temperature=0.99,
+    max_replay_buffer_size=20000,
+    min_replay_buffer_size=200,
+    log_interval=200,
+    min_learning_rate=1e-6,
+    lr_warmup_steps=1000,
+    lr_decay_steps=1000000,
+    epsilon_min=0.01,
+    epsilon_decay_steps=100000,
+    min_temperature=0.25,
+    temperature_decay_steps=100000,
+    s_weight=1e-2,
+    v_weight=0.1,
+    train_device="cpu",
+    actor_device="cpu",
+    reanalyze_device="cpu",
+    reanalyze_workers=0,
+)
+
+
+class MuZeroAgentAdapter(AgentAdapter):
+    def tensor_from_cog_obs(self, cog_obs, device=None):
+        pb_array = cog_obs.vectorized
+        np_array = np.frombuffer(pb_array.data, dtype=pb_array.dtype).reshape(*pb_array.shape)
+        return torch.tensor(np_array, dtype=self._dtype, device=device)
+
+    @staticmethod
+    def decode_cog_policy_value(cog_action):
+        policy = np.frombuffer(cog_action.policy.data, dtype=cog_action.policy.dtype).reshape(*cog_action.policy.shape)
+        value = cog_action.value
+        return policy, value
+
+    @staticmethod
+    def decode_cog_action(cog_action):
+        action = cog_action.discrete_action
+        return action
+
+    def __init__(self):
+        super().__init__()
+        self._model_cache = LRU(2)  # memory issue?
+        self._dtype = torch.float
+
+    def _create(self, model_id, *, obs_dim, act_dim, device, training_config):
+        return MuZeroAgent(obs_dim=obs_dim, act_dim=act_dim, device=device, training_config=training_config)
+
+    def _load(self, model_id, version_number, version_user_data, model_data_f):
+        return MuZeroAgent.load(model_data_f, "cpu")
+
+    def _save(self, model, model_data_f):
+        assert isinstance(model, MuZeroAgent)
+        model.save(model_data_f)
+        return {}
+
+    def _create_actor_implementations(self):
+        async def _single_agent_muzero_actor_implementation(actor_session):
+            actor_session.start()
+            agent, _ = await self.retrieve_version(actor_session.config.model_id, -1)
+            agent = copy.deepcopy(agent)
+            agent.set_device(actor_session.config.device)
```

Comment:
I'm not sure why this is part of the agent configuration, shouldn't it be tied to service ?

Reply:
Which part are you referring to? The deepcopy and set_device? There's actually a lot going on here that's a bit confusing. The actor implementation and run session implementation are in the same process and share the same model LRU cache through the AgentAdapter base class. This means that `retrieve_version` can return a reference to literally the same Python object that was created in the run session used for training. Which means that it can be modified in place during a trial, and might be on the wrong device (imagine running trials on CPU and training on GPU, or a multi-GPU setup). We work around this by deep copying the object and setting the device to the actor device (as oppose to the training device, which might be different).

Reply:
I was referring to the device setting, I think it should be a per-service setting not something that's decided per actor.

Regarding the deep copy I got it now, wouldn't it make sense to deep copy when we add to the LRU then ?

Reply:
Moving the deepcopy to LRU assumes that the model is deep copyable. If we're OK making this assumption, then I prefer to move it there.

Re: devices, we need to support different devices for trials and training. This needs to be configured somewhere. What alternative do you suggest?

Reply:
> Moving the deepcopy to LRU assumes that the model is deep copyable. If we're OK making this assumption, then I prefer to move it there.

Go for it it'll avoid a bunch of stupid problem

> Re: devices, we need to support different devices for trials and training. This needs to be configured somewhere. What alternative do you suggest?

an environment variable at the level of service. Or do you think there's a need for less granularity?

Reply:
Environment variable at the service level assumes we have a GPU configuration for training/trials/etc that is fixed once and for all by the service, that cannot be reconfigured for different runs. Why would we give up that flexibility? We agree that it needs to be configured somewhere, either through environment variables or in trial configuration. Why go for the less flexible option? E.g. service has access to 4 GPUs, why shouldn't it be up to the run configuration to decide which GPUs to use for which tasks? E.g. MuZero we could have 1 GPU for running trials, 1 for training, and 1 for reanalyze. Do you want to specify that in environment variables? Then we would have to spin off every agent/training algo into a separate service, and relaunch the services any time we change a configuration.

Reply:
Deepcopy has been removed as it was no longer required after other refactoring. The GPU configuration is something that we should leave for another PR, as it could affect other agent implementations as well.